### PR TITLE
Project UI support

### LINKED ISSUE #4132: make local UI work with projects
This is with the development binary release from https://github.com/unisonweb/unison/commit/daac38b1547649e1412976b8c88402595babf12b.

image

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

fixes #4132 

Makes the `ui` command and the codebase server project-aware in unison-local.

NOTE: This is a rudimentary first step, there isn't yet a branch browser or any way to switch between projects and loose code in the UI.

This works in conjunction with https://github.com/unisonweb/unison-local-ui/pull/12#pullrequestreview-1567778878 to provide a much nicer local project browsing experience.

**New project UI** (Great work @hojberg !)

image

**Old one:**

image

## Implementation notes

* Adds a new codeserver mount at `/projects/<project>/branches/<branch-ref>/*` which resolves the project branch into a root causal hash then calls through to the existing loose-code implementation.
* Updates the `ui` command to build the appropriate project URL when in a project.

## Test coverage

Tested locally against https://github.com/unisonweb/unison-local-ui/pull/12#pullrequestreview-1567778878

## Loose ends

~@hojberg  will need to add support to the UI for project-branch pages, and have it call the correct API endpoints from those pages.~ Apparently he was finished this long before I even started 😆

### CODE REVIEW

File: parser-typechecker/src/Unison/Project/Util.hs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+import U.Codebase.Sqlite.DbId (ProjectBranchId (..), ProjectId (..))
+import Unison.Codebase.Path qualified as Path
+import Unison.NameSegment (NameSegment (..))
+import Unison.Project (ProjectAndBranch (..))
+
+-- | Get the path that a project is stored at. Users aren't supposed to go here.
+--
+-- >>> projectPath "ABCD"
+-- .__projects._ABCD
+projectPath :: ProjectId -> Path.Absolute
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Project/Util.hs:24]: These were just moved out from CLI from `unison-cli/src/Unison/Cli/ProjectUtils.hs` so I can use them in the api server.

File: unison-cli/src/Unison/Cli/ProjectUtils.hs [hunk 2]
```diff
@@ -273,106 +275,3 @@ expectRemoteProjectBranchByTheseNames = \case
 remoteProjectBranchDoesntExist :: ProjectAndBranch ProjectName ProjectBranchName -> Cli void
 remoteProjectBranchDoesntExist projectAndBranch =
   Cli.returnEarly (Output.RemoteProjectBranchDoesntExist Share.hardCodedUri projectAndBranch)
-
-------------------------------------------------------------------------------------------------------------------------
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Cli/ProjectUtils.hs:277]: These were all just moved to `Unison.Project.Util` in parser-typechecker so they're available in the local ui server too.

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 3]
```diff
@@ -551,73 +546,7 @@ loop e = do
               Cli.updateRoot prev description
               (ppe, diff) <- diffHelper (Branch.head prev) (Branch.head rootBranch)
               Cli.respondNumbered (Output.ShowDiffAfterUndo ppe diff)
-            UiI path' -> do
-              Cli.Env {serverBaseUrl, codebase} <- ask
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:555]: Pulled this out into its own module.

**Comment by ChrisPenner** [comment]
Hold up a sec, found one small bug

**Comment by ChrisPenner** [comment]
Okay should be ready now, will need to co-ordinate with Simon when merging 👍🏼

**Review by hojberg — APPROVED** [review]
Sweet!!

**Pull request merged**